test/client-unit/services/StockEntryModalForm.spec.js

```diff
@@ -0,0 +1,184 @@
+/* global inject, expect, chai */
+/* eslint no-unused-expressions:off, no-restricted-properties:off */
+describe('StockEntryModalForm', () => {
+  beforeEach(module('bhima.services'));
+
+  let StockForm;
+
+  beforeEach(inject(_StockEntryModalForm_ => {
+    StockForm = _StockEntryModalForm_;
+  }));
+
+  const clone = obj => JSON.parse(JSON.stringify(obj));
+
+  it('#constructor() sets up rows as an empty array', () => {
+    const form = new StockForm();
+    expect(form.rows).to.be.a('array');
+    expect(form.rows).to.have.length(0);
+  });
+
+  it('#constructor() calls validate() if rows are passed in as options', () => {
+    const spy = chai.spy(StockForm.prototype.validate);
+    StockForm.prototype.validate = spy;
+
+    const form = new StockForm({
+      rows : [{}, {}],
+    });
+
+    expect(spy).to.have.been.called.exactly(1);
+    expect(form.rows).to.have.length(2);
+  });
+
+  it('#addItem() should add an item to the rows array', () => {
+    const form = new StockForm();
+    form.addItem();
+    expect(form.rows).to.have.length(1);
+    form.addItem();
+    expect(form.rows).to.have.length(2);
+  });
+
+  it('#removeItem() should remove an item from the rows array at an index', () => {
+    const form = new StockForm();
+    form.addItem();
+    form.addItem();
+    form.addItem();
+
+    expect(form.rows).to.have.length(3);
+
+    // mock some identities
+    form.rows.forEach((row, index) => {
+      row.id = index;
+    });
+
+    // remove the last item
+    form.removeItem(2);
+
+    let lastItem = form.rows[form.rows.length - 1];
+    expect(lastItem.id).to.equal(1);
+
+    form.removeItem(0);
+    lastItem = form.rows[form.rows.length - 1];
+    expect(lastItem.id).to.equal(1);
+  });
+
+  it('#total() calculates the totals of all quantities', () => {
+    const form = new StockForm({
+      rows : [{ quantity : 15 }, { quantity : 25 }],
+    });
+
+    const total = 40;
+    const calculated = form.total();
+    expect(total).to.equal(calculated);
+  });
+
+  const sampleRow = {
+    expiration_date : new Date(),
+    lot : 1234,
+    quantity : 10,
+  };
+
+  it('#validate() will set isValid to true for a valid row', () => {
+    const form = new StockForm({
+      rows : [sampleRow],
+    });
+
+    form.validate();
```

Comment:
This method `validate()` is called by the constructor of `StockForm` for predefined rows : 
```js
const form = new StockForm({
  rows : [sampleRow],
});
```
So, it is not necessary to call it again here.